### ISSUE TITLE
kvm2 driver validation times out after a libvirtd restart

### ISSUE THREAD

**Issue by macaptain** [kubernetes/minikube#11113]
**Steps to reproduce the issue:** 

On a CentOS or Ubuntu host, with libvirtd (version 6.0.0) managed by systemd:

Run: `sudo systemctl restart libvirtd; minikube start`

**Full output of failed command:** 

```
$ sudo systemctl restart libvirtd; minikube start
* minikube v1.19.0 on Ubuntu 20.04
* Using the kvm2 driver based on existing profile

X Exiting due to PROVIDER_KVM2_ERROR: /usr/bin/virsh domcapabilities --virttype kvm failed:

signal: killed
* Suggestion: Follow your Linux distribution instructions for configuring KVM
* Documentation: https://minikube.sigs.k8s.io/docs/reference/drivers/kvm2/
```

However, if I add a sleep command between the libvirtd restart and the `minikube start`, I'm on the happy path:

```
$ sudo systemctl restart libvirtd; sleep 5; minikube start
* minikube v1.19.0 on Ubuntu 20.04
* Using the kvm2 driver based on existing profile
* Downloading driver docker-machine-driver-kvm2:
    > docker-machine-driver-kvm2....: 65 B / 65 B [----------] 100.00% ? p/s 0s
    > docker-machine-driver-kvm2: 11.44 MiB / 11.44 MiB  100.00% 123.11 MiB p/s
* Downloading VM boot image ...
...
```

Note that the timeout in the execution context for running `/usr/bin/virsh domcapabilities --virttype kvm` is 2 seconds:

https://github.com/kubernetes/minikube/blob/0de446dc9ef1914ead7fac081c6ad7af83c167c5/pkg/minikube/registry/drvs/kvm2/kvm2.go#L116-L117

However, it takes longer than 2 seconds for `/usr/bin/virsh domcapabilities --virttype kvm` to exit on my system immediately after a libvirtd restart:

```
$ sudo systemctl restart libvirtd; time /usr/bin/virsh domcapabilities --virttype kvm >/dev/null

real    0m3.235s
user    0m0.022s
sys     0m0.009s
```

(without restarting libvirtd)
```
$ time /usr/bin/virsh domcapabilities --virttype kvm >/dev/null

real    0m0.054s
user    0m0.020s
sys     0m0.012s
```

**Comment by macaptain** [comment]
I'll raise a PR to bump the timeout to 5 seconds, if that's alright.

/assign

**Comment by afbjorklund** [comment]
For some of the other daemons (looking at you, containerd), we do a spin loop just to wait for a socket and a "I'm alive"

Maybe that can be done for the libvirt daemon as well. Or you could give it some cooldown after you restart it, of course.

**Comment by afbjorklund** [comment]
For instance, when I run this on the Raspberry Pi it will probably hit some other random time limit (2s, 5s, whatever)

**Comment by macaptain** [comment]
@afbjorklund that's a great suggestion. Ultimately, the timeout number will be arbitrary, but we'll be able to display a much better error message with your proposal. I'd be happy to change the PR so that just before we usually run the `virsh domcapabilities --virttype kvm`, we:

1. Try a `virsh connect` with a timeout context of, say, 10 seconds.
2. If `err != nil`, return a `registry.State` with an error message that the attempt to connect to the hypervisor timed out.
3. Proceed with the usual `virsh domcapabilities --virttype kvm` command with the same old timeout of 2 seconds (although, is 2 seconds a bit too aggressive? I guess it's OK as no one has complained or noticed up until now!)

Does that sound OK? Let me know if you think anything should be adjusted.

**Comment by afbjorklund** [comment]
That sounds perfectly reasonable. BTW, why did you have to restart libvirt in the first place ?

**Comment by afbjorklund** [comment]
Should also note that _normally_, a `virsh connext` goes very quickly (just a unix socket away)

```
time virsh connect


real	0m0,038s
user	0m0,023s
sys	0m0,015s
```

**Comment by macaptain** [comment]
> That sounds perfectly reasonable. BTW, why did you have to restart libvirt in the first place ?

I'm using Minikube in a CI pipeline. The `minikube start` is in a script that runs after some host setup, and among other things, libvirtd gets restarted there, so we started seeing this error flakily in CI runs. I'm not sure I have to restart libvirtd at this point, it looks to be more defensive in the CI script than necessary. But since I was able to reproduce this timeout consistently, I thought it was worth fixing. And it wasn't clear from the output that I was hitting a minikube timeout, so it will be a good thing to improve the error message for the future.
